app/init/draft_upload_manager.ts

```diff
@@ -0,0 +1,146 @@
+        file: FileInfo,
+        channelId: string,
+        rootId: string,
+        skipBytes = 0,
+    ) => {
+        this.handlers[file.clientId!] = {
+            fileInfo: file,
+            serverUrl,
+            channelId,
+            rootId,
+            lastTimeStored: 0,
+        };
+
+        const onProgress = (progress: number) => {
+            this.handleProgress(file.clientId!, progress);
+        };
+
+        const onComplete = (response: ClientResponse) => {
+            this.handleComplete(response, file.clientId!);
+        };
+
+        const onError = (response: ClientResponseError) => {
+            this.handleError(response.message, file.clientId!);
+        };
+
+        const {error, cancel} = uploadFile(serverUrl, file, channelId, onProgress, onComplete, onError, skipBytes);
+        if (error) {
+            this.handleError(error.message, file.clientId!);
+            return;
+        }
+        this.handlers[file.clientId!].cancel = cancel;
+    };
+
+    public cancel = (clientId: string) => {
+        const {cancel} = this.handlers[clientId];
+        delete this.handlers[clientId];
+        cancel?.();
+    };
+
+    public isLoading = (clientId: string) => {
+        return Boolean(this.handlers[clientId]);
+    };
+
+    private handleProgress = (clientId: string, progress: number) => {
+        const h = this.handlers[clientId];
+        if (!h) {
+            return;
+        }
+
+        h.fileInfo.progress = progress;
+
+        DeviceEventEmitter.emit(`${Events.FILE_PROGRESS}_${clientId}`, progress);
+        if (AppState.currentState !== 'active' && h.lastTimeStored + PROGRESS_TIME_TO_STORE < Date.now()) {
+            updateDraftFile(h.serverUrl, h.channelId, h.rootId, this.handlers[clientId].fileInfo);
+            h.lastTimeStored = Date.now();
+        }
+    };
+
+    private handleComplete = (response: ClientResponse, clientId: string) => {
+        const h = this.handlers[clientId];
+        if (!h) {
+            return;
+        }
+        if (response.code !== 201) {
+            this.handleError((response.data as any).message, clientId);
+            return;
+        }
+        if (!response.data) {
+            this.handleError('Failed to upload the file: no data received', clientId);
+            return;
+        }
+        const data = response.data.file_infos as FileInfo[];
+        if (!data || !data.length) {
+            this.handleError('Failed to upload the file: no data received', clientId);
+            return;
+        }
+
+        delete this.handlers[clientId!];
+
+        const fileInfo = data[0];
+        fileInfo.clientId = h.fileInfo.clientId;
+        fileInfo.localPath = h.fileInfo.localPath;
+
+        updateDraftFile(h.serverUrl, h.channelId, h.rootId, this.handlers[clientId].fileInfo);
+    };
+
+    private handleError = (errorMessage: string, clientId: string) => {
+        const h = this.handlers[clientId];
+        delete this.handlers[clientId!];
```

Comment:
Do you need the force unwrap on the `clientId` ?
```
        delete this.handlers[clientId!];
```

Reply:
Nope. Probably leftovers from different implementation :sweat_smile: